File: BaekJoon/ITBigTech/CrossCountry9017/CrossCountry.py
T = int(input())

for _ in range(T):
    N = int(input())
    table = list(map(int, input().split()))

    team_members = {}

    # 팀별 기록
    for i, team in enumerate(table):
        if team not in team_members:
            team_members[team] = []
        # 등수
        team_members[team].append(i+1) 

    # 6명 이상 팀만 필터링
    valid_teams = {team : ranks for team, ranks in team_members.items() if len(ranks) >= 6}
    
    # 유효한 팀별 개인 등수 정렬
    all_ranks = []
    for team, ranks in valid_teams.items():
        # pos 기준으로 sort를 하기 위해 pos를 앞으로
        all_ranks.extend((pos,team) for pos in ranks)
    all_ranks.sort()

    # 유효한 팀 대상 새로운 순위 및 점수
    score_list = {}
    for i, (pos,team) in enumerate(all_ranks):
        if team not in score_list:
            score_list[team] = []
        score_list[team].append(i+1)

    # 상위 4명 점수 및 우승팀
    min_score = float('inf')
    winner = -1
    for team, scores in score_list.items():
        team_score = sum(scores[:4])
        if team_score < min_score:
            min_score = team_score
            winner = team
        elif team_score == min_score:
            # 동점일 경우 5번째 비교
            if scores[4] < score_list[winner][4]:
                winner = team

    print(winner)

File: BaekJoon/ITBigTech/CrossCountry9017/CrossCountry.go
package main

import (
    "fmt"
    "strings"
    "strconv"
    "os"
    "bufio"
    "sort"
)

type Position struct {
    pos int
    team int
}

func main() {
    reader := bufio.NewReader(os.Stdin)
    T, _ := strconv.Atoi(readLine(reader))

    for i := 0; i < T; i++ {
        _ = readLine(reader) // 사용안해서 버림
        teams := strings.Split(readLine(reader), " ")

        teamMembers := make(map[int][]int)

        // 팀별 기록
        for idx, t := range teams {
            team, _ := strconv.Atoi(t)
            teamMembers[team] = append(teamMembers[team],idx+1)
        }

        validTeams := make(map[int][]int)
        for team, members := range teamMembers {
            if len(members) >= 6 {
                validTeams[team] = members
            }
        }

        // 유효한 팀 대상 새로운 순위 및 점수
        var allRanks []Position
        for team, members := range validTeams {
            for _, pos := range members {
                allRanks = append(allRanks, Position{pos, team})
            }
        }

        // pos 기준으로 정렬
        sort.Slice(allRanks, func(i, j int) bool {
            return allRanks[i].pos < allRanks[j].pos
        })

        scoreList := make(map[int][]int)
        for i, p := range allRanks {
            scoreList[p.team] = append(scoreList[p.team], i+1)
        }

        // 상위 4명 점수 계산 및 우승팀
        minScore := int(^uint(0) >> 1) // 최대 정수값
        winner := -1
        for team, scores := range scoreList {
            teamScore := sum(scores[:4])
            if teamScore < minScore {
                minScore = teamScore
                winner = team
            } else if teamScore == minScore {
                if scores[4] < scoreList[winner][4] {
                    winner = team
                }
            }
        }
        fmt.Println(winner)
    }
}

func readLine(reader *bufio.Reader) string {
    line, _ := reader.ReadString('\n')
    return strings.TrimSpace(line)
}

func sum(slice []int) int {
    total := 0
    for _, num := range slice {
        total += num
    }
    return total
}


